lib/src/embedded/compiler.ts

```diff
@@ -118,10 +84,132 @@ export class EmbeddedCompiler {
       .subscribe(() => process.nextTick(() => this.close()));
   }
 
+  /**
+   * Dispatches a compile request. Returns a promise that resolves when a
+   * compile response with a matching ID is received from the embedded compiler.
+   */
+  compile(request: InboundMessage.CompileRequest) {
+    const id = this.nextRequestId();
+    request.setId(id);
+    const message = new InboundMessage();
+    message.setCompilerequest(request);
+
+    this.recordRequest(id);
+    this.messageTransformer.write$.next(message);
+
+    return this.messageTransformer.read$
+      .pipe(
+        filter(message => {
+          return (
+            message.hasCompileresponse() &&
+            message.getCompileresponse()!.getId() === id
+          );
+        }),
+        map(message => message.getCompileresponse()!),
+        tap(response => this.recordResponse(response.getId())),
+        take(1),
+        takeUntil(this.error$)
+      )
+      .toPromise();
+  }
+
+  listen(
+    messageType: OutboundMessage.MessageCase,
+    processRequest: (request: Request) => Response
+  ) {
+    this.messageTransformer.read$
+      .pipe(
+        filter(message => message.getMessageCase() === messageType),
+        map(message => getRequest(message)),
+        tap(request => this.recordRequest(request.getId())),
+        map(request => processRequest(request)),
```

Comment:
Consider automatically adding the appropriate ID to the response object, so you don't need to duplicate that logic across each caller.